### PR TITLE
Add unit test recipe to the cookbook

### PR DESCRIPTION
**Pull request by chsienki**

### CODE REVIEW

**Comment by chsienki** [comment]
@dotnet/roslyn-compiler for doc review please :)

**Comment by 333fred** [comment]
Do we perhaps want to add an example of extracting the main logic out to a helper method?

File: docs/features/source-generators.cookbook.md [hunk 1]
```diff
@@ -620,6 +620,110 @@ public class MyGenerator : ISourceGenerator
 }
 ```
 
+### Unit Test generators
```
Comment (RikkiGibson) [inline docs/features/source-generators.cookbook.md:623]: maybe a dumb nit: consider titling this 'Unit testing of generators' so nobody gets the notion at a glance that this is about writing a source generator which generates unit test cases

File: docs/features/source-generators.cookbook.md [hunk 2]
```diff
@@ -620,6 +620,110 @@ public class MyGenerator : ISourceGenerator
...
+            Debug.Assert(generatorResult.Generator == generator);
+            Debug.Assert(generatorResult.Diagnostics.IsEmpty);
+            Debug.Assert(generatorResult.GeneratedSources.Length == 1);
+            Debug.Assert(generatorResult.Exception is null);
+        }
+
+        private static Compilation CreateCompilation(string source)
+            => CSharpCompilation.Create("compilation",
+                new[] { CSharpSyntaxTree.ParseText(source) },
+                new[] { MetadataReference.CreateFromFile(typeof(Binder).GetTypeInfo().Assembly.Location) },
```
Comment (RikkiGibson) [inline docs/features/source-generators.cookbook.md:719]: Doesn't Binder have internal accessibility? Unless Binder doesn't mean what I think it means here. It looks like this method of getting the Microsoft.CodeAnalysis.CSharp assembly reference won't work for external users? Maybe `typeof(CSharpCompilation)` would make more sense?

Comment (RikkiGibson) [reply docs/features/source-generators.cookbook.md:719]: realizing this is probably a System.Reflection.Binder, so moot. :smile:

**Review by RikkiGibson — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

File: docs/features/source-generators.cookbook.md (hunk 2 above)
Comment (chsienki) [reply docs/features/source-generators.cookbook.md:719]: Yeah, its actually just the easiest way to pull in all the right assemblies to make a compilation succeed.

**Pull request merged**